Core/src/Geometry/Portal.cpp

```diff
@@ -0,0 +1,306 @@
+  }
+}
+
+Portal::Portal(const GeometryContext& gctx, Arguments&& args) {
+  if (!args.alongNormal.m_surface && !args.oppositeNormal.m_surface) {
+    throw std::invalid_argument("At least one link must be provided");
+  }
+
+  if (args.alongNormal.m_surface) {
+    setLink(
+        gctx, Direction::AlongNormal,
+        std::make_unique<TrivialPortalLink>(
+            std::move(args.alongNormal.m_surface), *args.alongNormal.m_volume));
+  }
+  if (args.oppositeNormal.m_surface) {
+    setLink(gctx, Direction::OppositeNormal,
+            std::make_unique<TrivialPortalLink>(
+                std::move(args.oppositeNormal.m_surface),
+                *args.oppositeNormal.m_volume));
+  }
+}
+
+void Portal::setLink(const GeometryContext& gctx, Direction direction,
+                     std::unique_ptr<PortalLinkBase> link) {
+  if (link == nullptr) {
+    throw std::invalid_argument("Link must not be null");
+  }
+
+  auto& target =
+      direction == Direction::AlongNormal ? m_alongNormal : m_oppositeNormal;
+  const auto& other =
+      direction == Direction::AlongNormal ? m_oppositeNormal : m_alongNormal;
+
+  // check if surfaces are identical
+  if (m_surface != nullptr &&
+      !isSameSurface(gctx, link->surface(), *m_surface)) {
+    throw PortalFusingException();
+  }
+
+  // check if they both have material but are not the same surface
+  if (m_surface != nullptr && (m_surface.get() != &link->surface()) &&
+      link->surface().surfaceMaterial() != nullptr &&
+      m_surface->surfaceMaterial() != nullptr) {
+    throw PortalFusingException();
+  }
+
+  target = std::move(link);
+
+  if (other == nullptr) {
+    // We don't have an existing surface, take the one we just got
+    m_surface = target->surfacePtr();
+    return;
+  }
+
+  if (target->surface().surfaceMaterial() != nullptr) {
+    // new link has material: assign that to existing link
+    m_surface = target->surfacePtr();
+    other->setSurface(m_surface);
+  } else {
+    // none have material, or the existing surface had material: assign the
+    // existing surface by convention
+    target->setSurface(m_surface);
+  }
+}
+
+void Portal::setLink(const GeometryContext& gctx, Direction direction,
+                     std::shared_ptr<RegularSurface> surface,
+                     TrackingVolume& volume) {
+  setLink(gctx, direction,
+          std::make_unique<TrivialPortalLink>(std::move(surface), volume));
+}
+
+const PortalLinkBase* Portal::getLink(Direction direction) const {
+  if (direction == Direction::AlongNormal) {
+    return m_alongNormal.get();
+  } else {
+    return m_oppositeNormal.get();
+  }
+}
+
+Result<const TrackingVolume*> Portal::resolveVolume(
+    const GeometryContext& gctx, const Vector3& position,
+    const Vector3& direction) const {
+  assert(m_surface != nullptr);
+  const Vector3 normal = m_surface->normal(gctx, position);
+  Direction side = Direction::fromScalar(normal.dot(direction));
```

Comment:
I think this asserts on the parameter not being `0` - is that wanted here?

Reply:
Do you mean if `direction` and  the `normal` are perpendicular? I guess that's indeed an issue, we have no idea which volume to look up in that case.

Reply:
Changed to that.